corfudb-tools/src/main/java/org/corfudb/compactor/CorfuStoreCompactor.java

```diff
@@ -0,0 +1,313 @@
+                .sorted((Comparator<Map.Entry<?, Token>>) (o1, o2) -> o1.getValue().compareTo(o2.getValue()))
+                .map(entry -> new StringBuilder("{component: ")
+                        .append(entry.getKey())
+                        .append(", token: ")
+                        .append(entry.getValue())
+                        .append("}"))
+                .collect(Collectors.joining(", "));
+    }
+
+    void checkpoint() {
+
+        log.info("Starting check-pointing task");
+
+        CorfuStore corfuStore = new CorfuStore(corfuRuntime);
+
+        List<TableName> tableNames = new ArrayList<>(corfuStore.listTables(null));
+
+        ISerializer protobufSerializer = corfuRuntime.getSerializers().getSerializer(PROTOBUF_SERIALIZER_CODE);
+
+        try {
+            ISerializer keyDynamicProtobufSerializer = new KeyDynamicProtobufSerializer(corfuRuntime);
+            corfuRuntime.getSerializers().registerSerializer(keyDynamicProtobufSerializer);
+
+            // Measure time spent on check-pointing.
+            long startTime = System.nanoTime();
+
+            Token newToken = appendUfoCheckpoint(
+                    TableName.newBuilder()
+                            .setNamespace(TableRegistry.CORFU_SYSTEM_NAMESPACE)
+                            .setTableName(TableRegistry.REGISTRY_TABLE_NAME)
+                            .build(),
+                    keyDynamicProtobufSerializer,
+                    false);
+
+            if (enableTrim) {
+                // TODO: have a <K,V> type map to table name and loop/open/cp
+                // Needed by compactor
+                newToken = Token.min(
+                    appendCheckpoint(getCheckpointMap(), CorfuRuntimeHelper.CHECKPOINT),
+                    newToken
+                );
+                // Needed by compactor
+                newToken = Token.min(
+                    appendCheckpoint(getNodeTrimTokenMap(), CorfuRuntimeHelper.NODE_TOKEN),
+                    newToken
+                );
+            }
+
+            for (TableName tableName : tableNames) {
+                boolean diskBacked = persistedCacheRoot != null;
+                final Token trimToken = appendUfoCheckpoint(tableName,
+                        keyDynamicProtobufSerializer, diskBacked);
+                newToken = Token.min(newToken, trimToken);
+            }
+
+            long endTime = System.nanoTime();
+
+            for (int i = 0; i < txRetries; i++) {
+                try {
+                    try {
+                        corfuRuntime.getObjectsView().TXBegin();
+                        final String SpecialToken = "ufo";
+                        Token previousToken = getCheckpointMap().get(SpecialToken);
+                        if (previousToken == null || previousToken.getEpoch() <= newToken.getEpoch()
+                                && previousToken.getSequence() <= newToken.getSequence()) {
+                            getCheckpointMap().put(SpecialToken, newToken);
```

Comment:
Should the condition when previousToken and newToken are equal (equal sign) be skipped? As in if the previousToken had checkpointing done, can we skip it?